Add torch.float8_e5m2 and torch.float8_e4m3 data types

australopitek opened the pull request:
Proposal of two float8 variants - e5m2 and e4m3 - based on https://arxiv.org/pdf/2209.05433.pdf

Hide all Float8 operator implementations behind `#if !defined(C10_MOBILE)` guard to keep Android build size almost unchanged

TODO: 
 - Refactor duplicated code
 - Cleanup unbalanced pragma pop in dtype utils
 - Add native implementation on the CUDA size
 

Co-authored-by: Nikita Shulga <nshulga@meta.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang

australopitek committed:
Add torch.float8_e5m2 and torch.float8_e4m3 data types

vadimkantorov commented:
tangentially related: https://github.com/pytorch/pytorch/issues/69364

albanD reviewed (commented):
This looks amazing! Thanks a lot for sharing all this hard work to help!
I'll let @vkuzo pick it up from here!

vkuzo commented on c10/util/Float8_e4m3.h:
```diff
@@ -0,0 +1,247 @@
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+inline float fp8e4m3_to_fp32_value(uint8_t input) {
```
nice! I didn't read line by line yet, but high level question - what's your level of confidence on the correctness of these functions?  Have these been matched up with NVIDIA's casts or https://github.com/pytorch/FBGEMM/blob/277677039bae25b2570a73013b03bfaa9d2a523e/fbgemm_gpu/include/fbgemm_gpu/quantize_ops_utils.h ?

australopitek replied:
It was quite extensively tested against Gaudi2, which supports these variants natively. However, in next commits, I'll add comparative tests for all bits configurations against some reference implementation.
Can you guarantee that fbgemm implementation is fully aligned with the specification? Also, do you know if there is some open fp8_e5m2 reference implementation I could use for testing?

vkuzo replied:
the fbgemm implementation has been used extensively at Meta but it hasn't been matched up against real hardware afaik.

Your test cases look good, I'd say if future changes on numerics are needed that can be done in future PRs and not block this one.  LGTM.

vkuzo reviewed (commented):
this looks great and it's a lot more complete and battle tested than the previous time we tried this!  I had a couple of initial questions inline.

@australopitek , as @albanD mentioned we finally agreed to add these to core.  We owe you a more detailed review next week, but at a high level, would you be up for getting this PR to green CI / ready to land, or do you prefer that we take it as is and finish on our side?

australopitek commented on c10/core/ScalarType.h:
```diff
@@ -492,6 +550,8 @@ inline std::ostream& operator<<(
   _(half, Half) /* 0 */                \
-  _(bfloat16, BFloat16) /* 1 */
+  _(bfloat16, BFloat16) /* 1 */        \
+  _(float8_e5m2, Float8_e5m2) /* 2 */  \
```
You're right, this shouldn't be there. We were exploring possibilities to use fp8 with autocast, but finally decided to use custom solutions.
In fact, I see that AT_FORAUTOCAST_SCALAR_TYPES is not use anywhere, the only usage were removed quite long time ago. I will clean this during further work.

australopitek commented on c10/util/Float8_e4m3-inl.h:
```diff
@@ -0,0 +1,257 @@
+
+/// Arithmetic
+
+inline C10_HOST_DEVICE Float8_e4m3 operator+(const Float8_e4m3& a, const Float8_e4m3& b) {
```
This is just for completeness and simple tests. We use scaled float8 operations.

australopitek commented:
@vkuzo ,
I'll be happy to bring this PR to the end. I'll add more tests and fix current CI issues. Probably mainly alignment with coding guidelines/linters, since I didn't focus on them so far.
Waiting for the full review from you.

australopitek committed:
Cleanup and linters refactor

vkuzo reviewed (commented):
Overall this looks great to me and this fits into how we want to extend the ecosystem around float8.  Thanks @australopitek for getting this PR to a really good state!  I only have a couple of nit comments:

1. can we match MLIR (https://mlir.llvm.org/doxygen/classmlir_1_1FloatType.html) and XLA and go with `Float8_E4M3FN` instead of `Float8_E4M3`?
2. can we also get a review from  @ezyang or @albanD on the framework parts of the PR?

Using your current implementation of the casts as an initial reference LGTM, we can adjust as needed in future PRs.

australopitek commented:
@vkuzo ,
sure, I'll adjust naming to the MLIR. I'll try to push all changes and corrections this week.

ezyang commented on aten/src/ATen/NumericUtils.h:
```diff
@@ -62,6 +64,18 @@ inline C10_HOST_DEVICE bool _isnan(at::BFloat16 val) {
+template <typename T,
+         typename std::enable_if<std::is_same<T, at::Float8_e5m2>::value, int>::type = 0>
+inline C10_HOST_DEVICE bool _isnan(T val) {
+  return at::_isnan(static_cast<float>(val));
```
We probably will need to replace this with a primitive at some point as this is highly unlikely to be optimal.

australopitek replied:
Changed to helper method
```
inline C10_HOST_DEVICE bool Float8_e4m3fn::isnan() const {
  return (x & 0b01111111) == 0b01111111;
}
```

ezyang commented on c10/core/ScalarType.h:
```diff
@@ -425,6 +479,8 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   constexpr auto b1 = ScalarType::Bool;
   constexpr auto bf = ScalarType::BFloat16;
+  constexpr auto b8 = ScalarType::Float8_e5m2;
+  constexpr auto h8 = ScalarType::Float8_e4m3;
```
b8/h8 are funny names to call these lol.

australopitek replied:
Are there funny enough to be changed, or can I leave there?

ezyang commented on c10/core/ScalarType.h:
```diff
@@ -462,23 +518,25 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
+      /* q3 */ {ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud, ud},
+      /* bf */ {bf, bf, bf, bf, bf, f4, f4, f8, c4, c4, c8, bf, ud, ud, ud, bf, b8, h8},
+      /* b8 */ {b8, b8, b8, b8, b8, f4, f4, f8, c4, c4, c8, b8, ud, ud, ud, b8, b8, h8},
+      /* h8 */ {h8, h8, h8, h8, h8, f4, f4, f8, c4, c4, c8, h8, ud, ud, ud, h8, h8, h8},
```
This says that when you combine Float8_e5m2 and Float8_e4m3, it promotes to Float8_e4m3. This seems suspicious to me, as it's inconsistent with float16 and bfloat16 promotion, which promotes to float32 per the RFC at https://github.com/pytorch/pytorch/issues/43049 Additionally, JAX currently doesn't support promotion between these two types. It would be safer to disallow promotion to start and let folks figure out what they want to do later.

australopitek replied:
Done.

ezyang commented on c10/core/ScalarType.h:
```diff
@@ -50,7 +52,9 @@ namespace c10 {
-  _(c10::bits16, Bits16) /* 22 */
+  _(c10::bits16, Bits16) /* 22 */                        \
+  _(c10::Float8_e5m2, Float8_e5m2) /* 23 */              \
+  _(c10::Float8_e4m3, Float8_e4m3) /* 24 */
```
@vkuzo, folks pointed out to me on Twitter https://twitter.com/ezyang/status/1665399428809629696 that there are multiple e4m3 variants; GraphCore has its own separate NaN variant. In JAX, these variants are distinguished, see https://github.com/jax-ml/ml_dtypes#specifications-of-implemented-floating-point-formats for their naming convention. Since I don't think we're expecting people to actually use these dtype names directly, it seems... OK to keep inline with the LLVM/JAX naming conventions?

australopitek replied:
Done.

ezyang commented on c10/util/Float8_e4m3-inl.h:
```diff
@@ -0,0 +1,257 @@
+    return c10::Float8_e4m3(0x7F, c10::Float8_e4m3::from_bits());
+  }
+  static constexpr c10::Float8_e4m3 denorm_min() {
+    return c10::Float8_e4m3(0x01, c10::Float8_e4m3::from_bits());
```
I didn't carefully audit these

ezyang commented on torch/csrc/utils/byte_order.cpp:
```diff
@@ -69,6 +69,13 @@ static inline void swapBytes64(void* ptr) {
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  uint8_t output;
+  memcpy(&output, data, sizeof(uint8_t));
+  return output;
```
I just want to point out that this is completely unnecessary; as uint8 is byte sized you can just read it directly out of the pointer, no need to memcpy lol

australopitek replied:
Removed.

ezyang commented on test/quantization/core/experimental/test_float8.py:
```diff
@@ -0,0 +1,126 @@
+        compare_binary_with_decimal(*number, dtype)
+
+if __name__ == '__main__':
+    pytest.main([__file__])
```
don't forget to fix the missing eol

ezyang commented on c10/util/floating_point_utils.h:
```diff
@@ -0,0 +1,49 @@
+  } fp32 = {f};
+  return fp32.as_bits;
+#endif
+}
```
Code motion only right?

australopitek replied:
Yes.

ezyang commented:
The infra bits look solid. Most important comments are naming https://github.com/pytorch/pytorch/pull/104242/files/a2049dbf8f921c925b70bd3714c1f5dbcd8200ed#diff-177ba64a7713615ba52421a93517b9e332bcf80a6909259fcccf36cf5f3eddc7 and promotion https://github.com/pytorch/pytorch/pull/104242/files/a2049dbf8f921c925b70bd3714c1f5dbcd8200ed#diff-177ba64a7713615ba52421a93517b9e332bcf80a6909259fcccf36cf5f3eddc7

albanD reviewed (commented):
Looks great. A few extra comments but Ed pointed out the main things.

australopitek commented on c10/util/Float8_e4m3-inl.h:
```diff
@@ -0,0 +1,257 @@
+    return c10::Float8_e4m3(0x08, c10::Float8_e4m3::from_bits());
+  }
+  static constexpr c10::Float8_e4m3 lowest() {
+    return c10::Float8_e4m3(0xFE, c10::Float8_e4m3::from_bits());
```
@albanD, I think it's correct. Look at the definitions of min and lowest in std::numeric_limits:
From https://en.cppreference.com/w/cpp/types/numeric_limits/min: "For floating-point types with denormalization, min() returns the minimum positive normalized value."
minimum positive normal for e4m3fn is 0b00001000 = 0x08

From https://en.cppreference.com/w/cpp/types/numeric_limits/lowest: "Returns the lowest finite value representable by the numeric type T"
lowest finite value (negative) for e4m3fn is 0b11111110 = 0xFE

albanD replied:
Ho sorry, my bad. I realized that later and removed my other comment about this and forgot to delete this one.
It is correct indeed!
The min/lowest definitions are not what I expected haha

australopitek commented on aten/src/ATen/NumericUtils.h:
```diff
@@ -92,6 +106,14 @@ inline C10_HOST_DEVICE bool _isinf(at::BFloat16 val) {
+}
+
+inline C10_HOST_DEVICE bool _isinf(at::Float8_e4m3 val) {
+  return at::_isinf(static_cast<float>(val));
```
Nice catch, done :)

australopitek commented on test/quantization/core/experimental/test_float8.py:
```diff
@@ -0,0 +1,126 @@
+    masked = torch.bitwise_and(asInt_odded, mask)
+    return masked.view(dtype)*signs
+
+@pytest.mark.parametrize("dtype", [torch.float8_e5m2, torch.float8_e4m3])
```
Done.

australopitek commented on test/quantization/core/experimental/test_float8.py:
```diff
@@ -0,0 +1,126 @@
+# Owner(s): ["oncall: quantization"]
+
+import torch
+import pytest
```
Done.

australopitek commented on test/quantization/core/experimental/test_float8.py:
```diff
@@ -0,0 +1,126 @@
+def test_creation_with_zeros(dtype):
+    x = torch.zeros(8, dtype=torch.float)
+    x8 = torch.zeros(8, dtype=dtype)
+    assert torch.count_nonzero(x == x8) == x.numel()
```
Just first sanity test. I can remove it if it's too basic.

malfet commented on aten/src/ATen/Dispatch.h:
```diff
@@ -515,6 +531,73 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
+  AT_DISPATCH_CASE(SCALARTYPE5, __VA_ARGS__)          \
+  AT_DISPATCH_CASE(SCALARTYPE6, __VA_ARGS__)
+
+#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(    \
```
Anyone remembers, what was the original idea behind `ALL_TYPES_AND_COMPLEX_AND`? 
But perhaps there is a time to add `ABSOLUTELY_ALL_TYPES` and `ALL_TYPES_AND_ALL_FLOATS`? (as I can't imagine want wants to add a support for F8_E5M2 but not for  F8_E4M3)

vkuzo replied:
to clarify, this is not blocking the PR - we can figure out how to clean up these abstractions independently

australopitek committed:
CI fixes - added typeNames for float8 types and marked functions as C10_HOST_DEVICE

australopitek commented:
@vkuzo ,
I have pushed next commit with changes after review and linters corrections. Few CI jobs still fail with two kinds of error. First is https://github.com/pytorch/pytorch/actions/runs/5508059225/jobs/10038844710?pr=104242, which I have a solution for, second is https://github.com/pytorch/pytorch/actions/runs/5508059225/jobs/10038847090?pr=104242, which I think is not related to my change.

Since I'm not very fluent with GitHub, is there a way to run these failing jobs locally, without pushing yet another commit?

vkuzo commented:
@australopitek, this looks great! A Meta employee (myself or @malfet can help) will have to land this since there are Meta-only changes which are needed (unrelated to any discussion in this PR, just some internal code which needs to sync with all existing dtypes in PyTorch).  Please let us know what a good handoff point would be, CI is mostly green and we are happy to help with the remaining issues if you'd like - let us know!

australopitek commented:
@vkuzo , @malfet , I'd be glad if you helped with these last bits. From CI logs I see there are mainly two kinds of issues left, but I'm off next two weeks due to holidays. I'll have limited access to this account, but will be able to log in occasionally.
How/when can we do the handoff?

malfet commented:
@australopitek I can probably take over the PR or open the new one and keep you as an author

malfet commented:
Here is the build command that fails:
```
cd /var/lib/jenkins/workspace/build/caffe2 && /opt/cache/bin/sccache /opt/cache/bin/clang++ -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTORCH_ENABLE_LLVM -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -I/var/lib/jenkins/workspace/build/aten/src -I/var/lib/jenkins/workspace/aten/src -I/var/lib/jenkins/workspace/build -I/var/lib/jenkins/workspace -I/var/lib/jenkins/workspace/cmake/../third_party/benchmark/include -I/opt/llvm/include -I/var/lib/jenkins/workspace/third_party/onnx -I/var/lib/jenkins/workspace/build/third_party/onnx -I/var/lib/jenkins/workspace/third_party/foxi -I/var/lib/jenkins/workspace/build/third_party/foxi -I/var/lib/jenkins/workspace/build/include -I/var/lib/jenkins/workspace/build/caffe2/aten/src -I/var/lib/jenkins/workspace/aten/src/ATen/.. -I/var/lib/jenkins/workspace/third_party/miniz-2.1.0 -I/var/lib/jenkins/workspace/torch/csrc/api -I/var/lib/jenkins/workspace/torch/csrc/api/include -I/var/lib/jenkins/workspace/c10/.. -isystem /var/lib/jenkins/workspace/build/third_party/gloo -isystem /var/lib/jenkins/workspace/cmake/../third_party/gloo -isystem /var/lib/jenkins/workspace/cmake/../third_party/googletest/googlemock/include -isystem /var/lib/jenkins/workspace/cmake/../third_party/googletest/googletest/include -isystem /var/lib/jenkins/workspace/third_party/protobuf/src -isystem /opt/conda/envs/py_3.11/include -isystem /var/lib/jenkins/workspace/third_party/gemmlowp -isystem /var/lib/jenkins/workspace/third_party/neon2sse -isystem /var/lib/jenkins/workspace/third_party/XNNPACK/include -isystem /var/lib/jenkins/workspace/third_party/ittapi/include -isystem /var/lib/jenkins/workspace/cmake/../third_party/eigen -isystem /var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/third_party/oneDNN/include -isystem /var/lib/jenkins/workspace/third_party/ideep/include -isystem /var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/include -isystem /var/lib/jenkins/workspace/third_party/googletest/googletest/include -isystem /var/lib/jenkins/workspace/third_party/googletest/googletest -D_GLIBCXX_USE_CXX11_ABI=1 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=braced-scalar-init -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wvla-extension -Wnewline-eof -Winconsistent-missing-override -Winconsistent-missing-destructor-override -Wno-range-loop-analysis -Wno-pass-failed -Wno-error=pedantic -Wno-error=old-style-cast -Wno-error=inconsistent-missing-override -Wno-error=inconsistent-missing-destructor-override -Wconstant-conversion -Wno-invalid-partial-specialization -Wno-unused-private-field -Wno-missing-braces -Wunused-lambda-capture -Qunused-arguments -fcolor-diagnostics -faligned-new -Werror -fno-math-errno -fno-trapping-math -Werror=format -DHAVE_AVX512_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -std=gnu++17 -fPIE -DCAFFE2_USE_GLOO -DTH_HAVE_THREAD -Wno-unused-variable -pthread -MD -MT caffe2/CMakeFiles/kernel_function_legacy_test.dir/__/aten/src/ATen/core/boxing/impl/kernel_function_legacy_test.cpp.o -MF CMakeFiles/kernel_function_legacy_test.dir/__/aten/src/ATen/core/boxing/impl/kernel_function_legacy_test.cpp.o.d -o CMakeFiles/kernel_function_legacy_test.dir/__/aten/src/ATen/core/boxing/impl/kernel_function_legacy_test.cpp.o -c /var/lib/jenkins/workspace/aten/src/ATen/core/boxing/impl/kernel_function_legacy_test.cpp
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/boxing/impl/kernel_function_legacy_test.cpp:7:
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/boxing/impl/test_helpers.h:7:
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/dispatch/Dispatcher.h:4:
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/boxing/KernelFunction.h:5:
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/stack.h:5:
In file included from /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue.h:1497:
/var/lib/jenkins/workspace/aten/src/ATen/core/ivalue_inl.h:1930:10: error: 'generic_to<std::__cxx11::basic_string<char>, at::Tensor>' is deprecated: IValues based on std::unordered_map are slow and deprecated. Please use c10::Dict<K, V> instead. [-Werror,-Wdeprecated-declarations]
  return generic_to(std::move(*this), _fake_type<T>{});
```

australopitek commented:
Thank you @malfet . I hope there won't be much more CI issues.

malfet commented on c10/util/Float8_e4m3fn.h:
```diff
@@ -0,0 +1,249 @@
+  inline C10_HOST_DEVICE bool isnan() const;
+};
+
+#ifdef __clang__
+#pragma GCC diagnostic pop
+#endif
+
```
Err, pop without a push...
```suggestion
```

malfet commented on c10/util/Float8_e5m2.h:
```diff
@@ -0,0 +1,147 @@
+  inline C10_HOST_DEVICE bool isinf() const;
+};
+
+#ifdef __clang__
+#pragma GCC diagnostic pop
+#endif
+
```
Naughty, naughty...
```suggestion
```

malfet committed:
Remove unbalanced `GCC diagnostic pop`

malfet commented:
> It looks like all `clang` based builds are failing...

And the reason for the failure, is that both `Float8_e4m3fn.h` and `Float8_e5m2.h` contained unbalanced:
```cpp
#ifdef __clang__
#pragma GCC diagnostic pop
#endif
```
which resulted in test validating legacy behavior to fail..

malfet commented:
@pytorchbot rebase

malfet commented:
Ok, and the following is needed to fix `device`->`host` function calls (which is an error on ROCm and internal builds, I though we had the same on OSS, but obviously it's not the case right now):
```cpp
diff --git a/c10/util/Float8_e4m3fn.h b/c10/util/Float8_e4m3fn.h
--- a/c10/util/Float8_e4m3fn.h
+++ b/c10/util/Float8_e4m3fn.h
@@ -41,13 +41,6 @@
 #include <string>
 #include <utility>
 
-// Standard check for compiling CUDA with clang
-#if defined(__clang__) && defined(__CUDA__) && defined(__CUDA_ARCH__)
-#define C10_DEVICE_HOST_FUNCTION __device__ __host__
-#else
-#define C10_DEVICE_HOST_FUNCTION
-#endif
-
 #include <typeinfo> // operator typeid
 
 namespace c10 {
@@ -106,6 +99,8 @@
   unsigned long nonsign_bsr;
   _BitScanReverse(&nonsign_bsr, (unsigned long)nonsign);
   uint32_t renorm_shift = (uint32_t)nonsign_bsr ^ 31;
+#elif defined(__CUDA_ARCH__)
+  uint32_t renorm_shift = __clz(nonsign);
 #else
   uint32_t renorm_shift = __builtin_clz(nonsign);
 #endif
diff --git a/c10/util/Half.h b/c10/util/Half.h
--- a/c10/util/Half.h
+++ b/c10/util/Half.h
@@ -52,13 +52,6 @@
 #include <sycl/sycl.hpp> // for SYCL 2020
 #endif
 
-// Standard check for compiling CUDA with clang
-#if defined(__clang__) && defined(__CUDA__) && defined(__CUDA_ARCH__)
-#define C10_DEVICE_HOST_FUNCTION __device__ __host__
-#else
-#define C10_DEVICE_HOST_FUNCTION
-#endif
-
 #include <typeinfo> // operator typeid
 
 namespace c10 {
@@ -170,7 +165,7 @@
  * mode and no operations on denormals) floating-point operations and bitcasts
  * between integer and floating-point variables.
  */
-inline float fp16_ieee_to_fp32_value(uint16_t h) {
+C10_HOST_DEVICE inline float fp16_ieee_to_fp32_value(uint16_t h) {
   /*
    * Extend the half-precision floating-point number to 32 bits and shift to the
    * upper part of the 32-bit word:
diff --git a/c10/util/floating_point_utils.h b/c10/util/floating_point_utils.h
--- a/c10/util/floating_point_utils.h
+++ b/c10/util/floating_point_utils.h
@@ -2,18 +2,9 @@
 
 #include <cstdint>
 
-// Standard check for compiling CUDA with clang
-#if defined(__clang__) && defined(__CUDA__) && defined(__CUDA_ARCH__)
-#define C10_DEVICE_HOST_FUNCTION __device__ __host__
-#else
-#define C10_DEVICE_HOST_FUNCTION
-#endif
-
-namespace c10 {
-
-namespace detail {
+namespace c10::detail {
 
-C10_DEVICE_HOST_FUNCTION inline float fp32_from_bits(uint32_t w) {
+C10_HOST_DEVICE inline float fp32_from_bits(uint32_t w) {
 #if defined(__OPENCL_VERSION__)
   return as_float(w);
 #elif defined(__CUDA_ARCH__)
@@ -29,7 +20,7 @@
 #endif
 }
 
-C10_DEVICE_HOST_FUNCTION inline uint32_t fp32_to_bits(float f) {
+C10_HOST_DEVICE inline uint32_t fp32_to_bits(float f) {
 #if defined(__OPENCL_VERSION__)
   return as_uint(f);
 #elif defined(__CUDA_ARCH__)
@@ -45,5 +36,4 @@
 #endif
 }
 
-} // namespace detail
-} // namespace c10
+} // namespace c10::detail

```

malfet committed:
Avoid host_device calls

And remove unused defines

malfet committed:
Update Half.h

Remove unused defined
Make fp16_ieee_to_fp32_value a host device function

malfet committed:
Update floating_point_utils.h

- Get rid of confusing define and use `C10_HOST_DEVICE`
- Use nested namespace declaration

malfet committed:
Update Float8_e4m3fn.h

Move __CUDA_ARCH__ first to fix windows build

malfet committed:
Update Half.h

Mark fp16_ieee_to_fp32_value with C10_HOST_DEVICE

albanD reviewed (approved):
Sounds good!

lezcano commented:
In the near future, tt may be good to add these dtypes to the relevant OpInfos to get an understanding of what works and what doesn't.

malfet committed:
Disable Float8 ops for mobile builds

malfet commented:
Applied following to avoid binary size increase:
```
diff --git a/aten/src/ATen/native/Copy.cpp b/aten/src/ATen/native/Copy.cpp
index 52ddf161eb..b8008b5ba4 100644
--- a/aten/src/ATen/native/Copy.cpp
+++ b/aten/src/ATen/native/Copy.cpp
@@ -65,8 +65,12 @@ void copy_same_type_transpose_(Tensor& self, const Tensor& src) {
   // The code below is implemented with the assumption that sizes are equal
   TORCH_INTERNAL_ASSERT_DEBUG_ONLY(self.sizes().equals(src.sizes()));
 
+#if !defined(C10_MOBILE)
   AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(kHalf, kBool, kBFloat16, kFloat8_e5m2, kFloat8_e4m3fn,
                                           kComplexHalf, self.scalar_type(), "copy_", [&] {
+#else
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(kHalf, kBool, kBFloat16, kComplexHalf, self.scalar_type(), "copy_", [&] {
+#endif
     scalar_t* sp = src.data_ptr<scalar_t>();
     scalar_t* rp = self.data_ptr<scalar_t>();
     scalar_t* bp = buf.data_ptr<scalar_t>();
diff --git a/aten/src/ATen/native/LinearAlgebra.cpp b/aten/src/ATen/native/LinearAlgebra.cpp
index 845e19ed11..c940981fea 100644
--- a/aten/src/ATen/native/LinearAlgebra.cpp
+++ b/aten/src/ATen/native/LinearAlgebra.cpp
@@ -1438,7 +1438,11 @@ static void addmm_impl_cpu_(
 
   if(!dispatched) {
     // Apply BLAS routine
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBFloat16, kFloat8_e5m2, kFloat8_e4m3fn,
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16,
+#endif
         result.scalar_type(), "addmm_impl_cpu_",
         [&]{
           using opmath_t = at::opmath_type<scalar_t>;
diff --git a/aten/src/ATen/native/Scalar.cpp b/aten/src/ATen/native/Scalar.cpp
index bdcc6f96f1..788f201afe 100644
--- a/aten/src/ATen/native/Scalar.cpp
+++ b/aten/src/ATen/native/Scalar.cpp
@@ -30,8 +30,13 @@ Scalar item(const Tensor& self) {
 
 Scalar _local_scalar_dense_cpu(const Tensor& self) {
   Scalar r;
+#if !defined(C10_MOBILE)
   AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(
     kComplexHalf, kHalf, kBool, kBFloat16, kFloat8_e5m2, kFloat8_e4m3fn, self.scalar_type(), "_local_scalar_dense_cpu", [&] {
+#else
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+    kComplexHalf, kHalf, kBool, kBFloat16, self.scalar_type(), "_local_scalar_dense_cpu", [&] {
+#endif
         scalar_t value = *self.data_ptr<scalar_t>();
         r = Scalar(value);
       });
diff --git a/aten/src/ATen/native/TensorCompare.cpp b/aten/src/ATen/native/TensorCompare.cpp
index d05f4e5562..ea30b0b07d 100644
--- a/aten/src/ATen/native/TensorCompare.cpp
+++ b/aten/src/ATen/native/TensorCompare.cpp
@@ -381,7 +381,11 @@ Tensor isinf(const Tensor &self) {
           (at::isinf(at::imag(self)));
   }
 
+#if !defined(C10_MOBILE)
   return AT_DISPATCH_FLOATING_TYPES_AND3(kBFloat16, kHalf, kFloat8_e5m2, self.scalar_type(), "isinf", [&]() {
+#else
+  return AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, self.scalar_type(), "isinf", [&]() {
+#endif
     return self.abs() == std::numeric_limits<scalar_t>::infinity();
   });
 }
@@ -397,7 +401,11 @@ Tensor isfinite(const Tensor& self) {
     return at::isfinite(at::real(self)).__iand__(at::isfinite(at::imag(self)));
   }
 
+#if !defined(C10_MOBILE)
   return AT_DISPATCH_FLOATING_TYPES_AND3(kHalf, kBFloat16, kFloat8_e5m2, self.scalar_type(), "isfinite", [&]() {
+#else
+  return AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, self.scalar_type(), "isfinite", [&]() {
+#endif
     return (self == self) * (self.abs() != std::numeric_limits<scalar_t>::infinity());
   });
 }
diff --git a/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp b/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
index fc8e749e18..0b8c28832f 100644
--- a/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
+++ b/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
@@ -85,7 +85,11 @@ void mul_kernel(TensorIteratorBase& iter) {
         });
     });
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, dtype, "mul_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(kBFloat16, kHalf, dtype, "mul_cpu", [&]() {
+#endif
       cpu_kernel_vec(iter,
         [=](scalar_t a, scalar_t b) __ubsan_ignore_undefined__ -> scalar_t { return a * b; },
         [=](Vectorized<scalar_t> a, Vectorized<scalar_t> b) __ubsan_ignore_undefined__ {
@@ -528,14 +532,22 @@ void ge_kernel(TensorIteratorBase& iter) {
 void eq_kernel(TensorIteratorBase& iter) {
   // See Note [special-case bool outputs]
   if (iter.dtype() == ScalarType::Bool) {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(kComplexHalf, kBool, kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, iter.common_dtype(), "eq_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(kComplexHalf, kBool, kBFloat16, kHalf, iter.common_dtype(), "eq_cpu", [&]() {
+#endif
       cpu_kernel(iter,
         [](scalar_t a, scalar_t b) -> bool {
           return a == b;
         });
     });
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND5(kComplexHalf, kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, iter.common_dtype(), "eq_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kComplexHalf, kBFloat16, kHalf, iter.common_dtype(), "eq_cpu", [&]() {
+#endif
       cpu_kernel_vec(
         iter,
         [](scalar_t a, scalar_t b) -> scalar_t {
@@ -551,14 +563,22 @@ void eq_kernel(TensorIteratorBase& iter) {
 void ne_kernel(TensorIteratorBase& iter) {
   // See Note [special-case bool outputs]
   if (iter.dtype() == ScalarType::Bool) {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(kComplexHalf, kBool, kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, iter.common_dtype(), "ne_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(kComplexHalf, kBool, kBFloat16, kHalf, iter.common_dtype(), "ne_cpu", [&]() {
+#endif
       cpu_kernel(iter,
         [](scalar_t a, scalar_t b) -> bool {
           return a != b;
         });
     });
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND5(kComplexHalf, kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, iter.common_dtype(), "ne_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kComplexHalf, kBFloat16, kHalf, iter.common_dtype(), "ne_cpu", [&]() {
+#endif
       cpu_kernel_vec(
         iter,
         [](scalar_t a, scalar_t b) -> scalar_t {
diff --git a/aten/src/ATen/native/cpu/CopyKernel.cpp b/aten/src/ATen/native/cpu/CopyKernel.cpp
index b83a62246b..ef0739f6bf 100644
--- a/aten/src/ATen/native/cpu/CopyKernel.cpp
+++ b/aten/src/ATen/native/cpu/CopyKernel.cpp
@@ -183,8 +183,13 @@ void direct_copy_kernel(TensorIteratorBase &iter) {
   } else if (dtype == ScalarType::ComplexHalf) {
     cpu_kernel(iter, [=](c10::complex<at::Half> a) -> c10::complex<at::Half> { return a; });
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND5(
         kBool, kHalf, kBFloat16, kFloat8_e5m2, kFloat8_e4m3fn, dtype, "copy_kernel", [&] {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+        kBool, kHalf, kBFloat16, dtype, "copy_kernel", [&] {
+#endif
       cpu_kernel_vec(
           iter,
           [=](scalar_t a) -> scalar_t { return a; },
@@ -237,9 +242,15 @@ void copy_kernel(TensorIterator& iter, bool /*non_blocking*/) {
     sizeof(BFloat16) == strides_out[0] && (sizeof(float) == strides_in[0] || strides_in[0] == 0)))) {
     float_bfloat16_copy_kernel(iter, requires_neg);
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, ScalarType::Float8_e5m2, ScalarType::Float8_e4m3fn, dtype, "copy_", [&] {
       using dest_t = scalar_t;
       AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND6(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, ScalarType::Float8_e5m2, ScalarType::Float8_e4m3fn, iter.dtype(1), "copy_", [&] {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, dtype, "copy_", [&] {
+      using dest_t = scalar_t;
+      AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(ScalarType::ComplexHalf, ScalarType::Half, ScalarType::Bool, ScalarType::BFloat16, iter.dtype(1), "copy_", [&] {
+#endif
         if (iter.has_contiguous_first_dim()) {
           TORCH_INTERNAL_ASSERT(iter.ninputs() == 1);
           TORCH_INTERNAL_ASSERT(iter.noutputs() == 1);
diff --git a/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp b/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp
index d7c79d4689..3109ac7317 100644
--- a/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp
+++ b/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp
@@ -186,7 +186,11 @@ static void abs_kernel(TensorIteratorBase& iter) {
     using opmath_t = at::opmath_type<scalar_t>;
     cpu_kernel(iter, [=](scalar_t a) -> scalar_t { return abs_impl(opmath_t{a}); });
   } else {
+#if !defined(C10_MOBILE)
     AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(kBFloat16, kHalf, kFloat8_e5m2, kFloat8_e4m3fn, iter.dtype(), "abs_cpu", [&]() {
+#else
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(kBFloat16, kHalf, iter.dtype(), "abs_cpu", [&]() {
+#endif
       cpu_kernel_vec(
           iter,
           [=](scalar_t a) -> scalar_t { return abs_impl(a); },
```

yuanyao-nv commented:
I see there have been some questions about agreement with Nvidia formats. Just wanted to bring up this info here: https://onnx.ai/onnx/technical/float8.html
Currently the ONNX spec for `E4M3FN` and `E5M2` matches Nvidia's FP8 formats. This page has detailed information on the representation of these two types and their rounding/saturation behavior.

malfet commented:
And now I need to figure out why MSVC preprocessor does not want to parse those ifdefs

malfet committed:
Will that help windowsw
:q
d

malfet committed:
More kernels

malfet committed:
And remaining kernels

malfet committed:
Fix tensor compare

malfet committed:
Fix another typo

malfet committed:
Oops

PaliC commented:
@pytorchbot revert -m "breaks lint (run lintrunner and remerge)" -i

PaliC commented:
@pytorchbot revert -m "breaks lint (run lintrunner and remerge)" -c ignoredsignal

PaliC commented:
@pytorchbot merge -f "already landed internally"

yuanyao-nv commented on c10/util/Float8_e5m2.h:
```diff
@@ -0,0 +1,143 @@
+
+  if (f_bits >= fp8_max) {
+    // NaN - all exponent and mantissa bits set to 1
+    result = f_bits > fp32_inf ? UINT8_C(0x7F) : UINT8_C(0x7C);
```
Can I ask if PyTorch is only implementing the non-saturation mode for casts, as specified [here ](https://onnx.ai/onnx/technical/float8.html#cast)? 

The current default behavior for ONNX is to saturate: https://github.com/onnx/onnx/blob/main/docs/Operators.md#attributes-7.

australopitek replied:
You're right, currently it implements only non-saturation mode. And, in consequence, is not aligned with float8 implementation from FBGEMM https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/include/fbgemm_gpu/quantize_ops_utils.h

@vkuzo, @albanD, was it intended to have saturation enabled by default in casting to float8 dtypes? If so, it needs to be added, because currently values out of float8 range are being casted to inf/nan. I can do it next week after vacation.

Also, should it be configurable via parameter of .to() operator or some other way?

vkuzo replied:
Thanks for the comment!  During PR review we punted these questions to future PRs just to get the initial dtype in, I guess now it's the time to discuss them :)

I would make an educated guess that saturating by default would make sense if people are doing delayed scaling, but let me think on that / talk to some of the research teams using fp8.  

For if/how to make this configurable...let me also ask around on what would make sense.

albanD replied:
In terms of consistency, when given integer inputs, we raise an error for too large numbers.
For floating point inputs, all our dtypes will return inf.

So I think inf is the right call here.

I think that `a_half = a.clamp(max=torch.finfo(torch.half).max).to(dtype=torch.half)` is a good way to get the behavior you want if needed.

yuanyao-nv replied:
Thanks for the information. 

Let me also add that when we introduced FP8 into ONNX one of the motivations for having saturation mode as the default was that during inference we don't want to end up with a lot of NaN's and Inf's propagating through the network.

cc: @nzmora @xadupre

vkuzo replied:
> Agreed that this might change when we have a stable decision on how to handle delayed scaling.
In the meantime keeping a behavior consistent with our other floating point dtypes (as done in this PR) sounds like the right way to go.

That sounds reasonable to me.  We'll just have to figure out how to expose the saturation option.  Ideally there should be a way to do to the saturated casts in eager mode without requiring a compiler pass.

australopitek replied:
@vkuzo ,
Is there an open discussion about saturation mode?

vkuzo replied:
Here is the current plan:
1. the default will stay unsaturated as is, to align with @albanD's thoughts on matching default behavior of other floating point dtypes
2. @malfet is working on a private API to expose the saturated casts, so that we can bind to hardware accelerated saturated casts without requiring graph capture -> graph pass.  Name of this is TBD.
3. long term, we may revisit this after we get more data on how people use these dtypes

vkuzo replied:
It will be exposed to Python with a name preceded by an underscore, but there will initially be no BC guarantees.

australopitek replied:
Thank you for the information.

yanbing-j replied:
Hi @vkuzo @albanD @australopitek ,

May I know the current status of enabling fp8 delayed scaling and saturation in PyTorch?

For fp8 delayed scaling, it is a per-tensor one. So, is the implementation similar with Quantized operators? It cannot just use CPU dispatch, right? Because scaling factor need to be stored in Tensor. Please correct me if anything wrong.

And you mentioned that fp8 training with delayed scaling, what about fp8 inference? In https://github.com/NVIDIA/TransformerEngine, fp8 inference is also with delayed scaling as training. Will you take this into consideration of fp8 enabling?

Thanks so much! Hope to hear from you guys!

cc @mingfeima

vkuzo replied:
Hi @yanbing-j , please feel free to message @gchanan on PyTorch slack for context on your question.  I am on parental leave until early December but Greg can help on this while I'm out.

malfet commented:
> @pytorchbot revert -m "breaks lint (run lintrunner and remerge)" -c ignoredsignal

Not sure I understand why you've added `ignoredsignal` category here, as lint on PR CI was green. Perhaps `landrace` would be a more appropriate category.

australopitek commented on c10/util/Float8_e4m3fn.h:
```diff
@@ -0,0 +1,240 @@
+ * Convert a 32-bit floating-point number in IEEE single-precision format to a
+ * 8-bit floating-point number in fp8 E4M3FN format, in bit representation.
+ */
+inline C10_HOST_DEVICE uint8_t fp8e4m3fn_from_fp32_value(float f) {
```
@yanbing-j it was based on the float16 datatype implementation https://github.com/pytorch/pytorch/blob/main/c10/util/Half.h with alignment to float8 bits configuration.

yuanyao-nv replied:
@yanbing-j Also see the casting rules tabulated here: https://onnx.ai/onnx/technical/float8.html

yanbing-j replied:
> @yanbing-j it was based on the float16 datatype implementation https://github.com/pytorch/pytorch/blob/main/c10/util/Half.h with alignment to float8 bits configuration.

What about the fp8_max? why we need to support special values here?

yanbing-j replied:
> @yanbing-j Also see the casting rules tabulated here: https://onnx.ai/onnx/technical/float8.html

Thanks. What is the value of FLT_MAX, does this support special values (in Section 3.1 in https://arxiv.org/pdf/2209.05433.pdf)?

australopitek replied:
@yanbing-j,
Part with fp8_max was probably inspired by Eigen implementation https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/arch/Default/Half.h#L625. Now I don't remember why I chose this one, but I guess it's analogous to PyTorch's implementation.

Special values are implemented according to mentioned paper and they're tested in https://github.com/pytorch/pytorch/blob/main/test/quantization/core/experimental/test_float8.py#L117 for both dtypes.

yanbing-j replied:
@australopitek Thank you so much. It really helps.
